Replaced config map usage with, omaps and volume ID generation

ShyamsundarR opened the pull request:
Existing config maps are now replaced with rados omaps that help
store information regarding the requested volume names and the rbd
image names backing the same.

Further to detect cluster, pool and which image a volume ID refers
to, changes to volume ID encoding has been done as per provided
design specification in the stateless ceph-csi proposal.

Signed-off-by: ShyamsundarR <srangana@redhat.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ceph/ceph-csi/288)
<!-- Reviewable:end -->

ShyamsundarR commented:
Testing done includes,
- Create delete PVC
- Create pods using PVC
- Create/promote/use/delete snapshots
- Create/delete 5 PVCs

TODO items are as follows:
- Update documentation across the repository
- Changes to deployment files
- Changes to examples directory

Tech Debt left for later (i.e optimizations and cleanup that does not hamper functionality and are not going to be a part of this PR):
- Create a cache for pool IDs, instead of fetching them each time
- Create a cache for secrets and move into watch mode for the same
- Unit testing cephcmds.go
- Fetch omaps in a list than one by one
- Provide tools to inspect omaps against the Ceph cluster

liewegas commented:
It wasn't possible to get away with a stateless / deterministic mapping from PV[C] to rbd image name?

ShyamsundarR commented:
> It wasn't possible to get away with a stateless / deterministic mapping from PV[C] to rbd image name?

Direct mapping of PVC (VolumeName) to rbd image name is feasible, but encoding cluster/pool information into the returned name (VolumeID) was breaking the spec rules, which states both VolumeName and VolumeID are 128 bytes maximum length. Thus, leaving no space in the VolumeID for cluster+pool IDs, if we encode the PVC name as well into the VolumeID.

In practice, kubernetes CSI implementation is using a PVC naming format of "pvc-[uuid]" thus leaving enough space to encode further information in the PVC name itself and return it as the VolumeID, but if the practice changes for any reason to use up more bytes in the VolumeName, it may become infeasible to use the same approach in the future.

The other (earlier) deterministic mapping using (say) xxhash of PVC name + instance ID for hash collisions, avoided the omap usage, but also required all images, in a pool, to be listed and loaded into in-memory maps by the CSI plugin, to ensure collisions are encoded by the right instance ID (including dealing with holes in the instance ID generation).

This hash+instanceID method was less elegant than using omaps to store the same, as the omap scheme does not need any form of instance ID generation and also did not require reading in all image names from a pool before provisioning a new instance on the same pool.

Stretch possibility: One way out of using any omap or hash scheme would be to get the CSI spec to change its sizing rules for the VolumeID, and say make that double of the VolumeName, which gives enough head room for encoding of a handle to use for internal purposes, but that ship may have sailed!

ShyamsundarR commented:
@gman0 @rootfs @dillaman @Madhu-1 PTAL.

It is a fat patch, I was unable to break it into 2 and still retain a mid-way between VolID encoding first and then removing the config maps. Do keep the feedback flowing, as you review it, to help with the direction or shepherding the patch towards an mergable state.

Madhu-1 commented on pkg/rbd/controllerserver.go:
```diff
@@ -130,76 +99,69 @@ func parseVolCreateRequest(req *csi.CreateVolumeRequest) (*rbdVolume, error) {
+		return &csi.CreateVolumeResponse{
+			Volume: &csi.Volume{
+				VolumeId:      rbdVol.VolID,
+				CapacityBytes: rbdVol.VolSize,
```
need to convert volume size back to bytes

Madhu-1 commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,9 +224,430 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	var vi util.VolumeIdentifier
+
+	// generate a uuid for the snap image name
+	uuid := uuid.NewUUID().String()
```
can we name this variable `uuid` to something else, because package name imported is also `uuid`

ShyamsundarR replied:
Will address in a future update to the PR.

Madhu-1 commented on pkg/util/cephcmds.go:
```diff
@@ -0,0 +1,364 @@
+
+	for _, p := range pools {
+		if poolName == p.Name {
+			return uint32(p.Number), nil
```
question: can we make struct filed as unit32 to avoid type conversion?

ShyamsundarR replied:
Do you mean the `Number` field at line 51? I think so, -ve pool IDs should not be valid anyway, so int should not be required there.

dillaman replied:
Correct, negative pool numbers are not valid (but -1 is used internally in the APIs to indicate pool-not-found / pool-not-valid versus the "0" returned below in the error case which is a possibly valid pool id).

ShyamsundarR replied:
So, making that uint32 would be fine, right?

Also, the line below has to return a value (in this case 0), but we should possibly rely on the go idiom to check errors before using return values, and hence the 0 below should be fine?

dillaman replied:
It's actually a ```int64_t``` internally (not that I expect 63bits of a pool number) [1]. 

[1] https://github.com/ceph/ceph/blob/master/src/include/rados/librados.h#L839

Madhu-1 commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,9 +224,430 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	poolID, err := util.GetPoolID(rbdSnap.Monitors, rbdSnap.AdminID, key,
+		rbdSnap.Pool)
+	if err != nil {
+		klog.V(4).Infof("Error fetching pool ID (%s)", err)
```
klog.Errorf() instead of info

ShyamsundarR replied:
Will address in a future update to the PR.

Madhu-1 commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,9 +224,430 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+func checkRBDSnapExists(rbdSnap *rbdSnapshot, credentials map[string]string) (found bool, err error) {
+	// Structure members used to determine if provided rbdSnapshot exists, are checked here, to
+	// provide an easy way to check when this function can be reused
+	if rbdSnap.RequestName == "" || rbdSnap.Monitors == "" || rbdSnap.AdminID == "" ||
```
do we need this validation here, I think this validation is already done in genRBDVolFromVolID and validateSnapshotReq

ShyamsundarR replied:
The comment above provides the context on why I added these checks.

In the future when thinking about reusing this function, the requirements for rbdVol need to be met, hence added this check as a future protection against reuse without a properly filled up rbdVol structure.

dillaman commented on pkg/rbd/rbd.go:
```diff
@@ -27,10 +27,89 @@ import (
+  - Key value contains the RBD snapshot name that is created or will be created, for the CO provided name
+
+- A per image omap named "csi.volume."+[RBD image name], we refer to this as the imgOMap
+  - stores a single key named "csi.volname", that has the value of the CO generated VolName that
```
The ```ListSnapshotsRequest``` CSI method has an optional ```SourceVolumeId``` field [1]. I assume that this would need to be a volume ID from this CSI, so should this also include a collection of snapshot ids? When combined w/ the comment below about ensuring that this object doesn't already exist, you can keep this object in the pool until both the volume and all associated snapshots have been removed if that ```SourceVolumeId``` link should remain valid even after the volume is removed.

[1] https://github.com/container-storage-interface/spec/blob/master/lib/go/csi/csi.pb.go#L3215

ShyamsundarR replied:
Yes, the assumption is correct, `ListSnapshotRequest` can come with a `SourceVolumeID` that is a request to only list snapshots of that source volume.

I was going to list all snapshots of the provided `SourveVolumeID` and filter out those created by the CSI plugin, and return the same in such a request (when implementing `ListSnapshots` capability i.e).

The above manner of listing assumes that the RBD image that backs the SourceVolumeID exists, when such a request comes. To overcome the possibility that we will support source image deletion if it has snapshots, the snapshot names need to be preserved in the `csi.volume.<image-uuid>` omap as suggested.

Currently though we do not allow delete of an image if it has snapshots, and way forward is being discussed in other issues and PRs.

What I propose for this PR is,
- Postpone stashing snapshot names in the volume omap and the process of retaining the omap until the list of keys within it is empty, till we decide on a course for the delete
OR,
- Just to ensure data is already stored on disk and prevent future incompatibility, add the snapshot name into the `csi.volume.<image-uuid>` omap, even if this is unused, but do not deal with retaining the omap on delete of the source image, which can be plumbed in later once the delete of source image with snapshots is considered.

Thoughts?

dillaman replied:
I guess we can avoid storing it for now and when support is added for deleting volumes w/ snapshots, the necessary lookups can be written to RADOS when the volume deletion request is processed.

ShyamsundarR replied:
Updated [comment](https://github.com/ceph/ceph-csi/pull/288#issuecomment-476837855) tracking TODO and TechDebt with List support line item, including a note on using the image omap to store the snapshot names.

dillaman commented on pkg/rbd/rbd_util.go:
```diff
@@ -53,21 +63,30 @@ type rbdVolume struct {
-	ClusterID          string `json:"clusterId"`
+	// SourceVolumeID is the volume ID of VolName, that is exchanged with CSI drivers
+	SourceVolumeID string `json:"sourceVolumeID"`
+	// VolName is the name of the RBD image, that is this rbdSnapshot's source image
```
Nit: confusing -- ```VolName``` elsewhere is the CSI volume name not the RBD image name. Rename to ```ImageName``` if that's it's intended use?

dillaman commented on pkg/rbd/rbd_util.go:
```diff
@@ -53,21 +63,30 @@ type rbdVolume struct {
+	// VolName is the name of the RBD image, that is this rbdSnapshot's source image
+	VolName string `json:"volName"`
+	// SnapName is the name of the RBD snapshot backing this rbdSnapshot
+	SnapName string `json:"snapName"`
```
Nit: also confusing if this is the RBD-internal snapshot name

dillaman commented on pkg/rbd/rbd_util.go:
```diff
@@ -126,7 +145,7 @@ func getMon(pOpts *rbdVolume, credentials map[string]string) (string, error) {
 
-// CreateImage creates a new ceph image with provision and volume options.
+// createRBDImage creates a new ceph image with provision and volume options.
 func createRBDImage(pOpts *rbdVolume, volSz int, adminID string, credentials map[string]string) error {
```
Nit: comment applies throughout: why add ```RBD``` to all these method names? They are inside the ```rbd``` package.

ShyamsundarR replied:
Truthfully all names were 'extended' from the existing function name createRBDExists.


Comment does make sense, but somewhere is also jarring (the RBD in there provide a sense of comfort that the RBD function is being called). Will relook and see what would be the right naming for these.

dillaman commented on pkg/rbd/rbd_util.go:
```diff
@@ -39,8 +45,12 @@ const (
 	rbdDefaultMounter        = "rbd"
 )
 
+// TODO: The following do not come from JSON anymore,  remove the json tags
```
Does this TODO apply to ```rbdSnapshot``` as well?

dillaman commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,9 +224,430 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+
+	rbdSnap.SnapName = snapName
+
+	// TODO: use listomapvals to dump all keys instead of reading them one-by-one
```
Nit: why would listing all keys be more efficient than retrieving a couple keys?

ShyamsundarR replied:
My thinking is as follows,
Instead of this container making 2 rados calls across the network to fetch key 1 and then 2, if it sends a list call, it saves the entire network round trip for the second fetch call.

Internally RADOS may already have cached the values for the keys, so I am not thinking along cost amortization on disk or media access.

dillaman replied:
Fair enough. Long term efficiency, we really should just get the entire CSI off using the CLI (and associated parsing of CLI output and error codes) and instead directly use the golang wrapper API [1]

[1] https://github.com/ceph/go-ceph

dillaman commented on pkg/rbd/rbd_util.go:
```diff
@@ -30,6 +35,7 @@ import (
 
 const (
 	imageWatcherStr = "watcher="
+	rbdImageFormat1 = "1"
```
Image format v1 is long-since deprecated -- don't allow users to attempt to create them since it will just fail.

dillaman commented on pkg/util/volid.go:
```diff
@@ -0,0 +1,160 @@
+  ability to parse backward compatible encodings.
+- ClusterID: Is a unique ID per cluster that the CSI instance is serving and is restricted to
+  lengths that can be accommodated in the encoding scheme.
+- ImageName: Is the on-disk image name of the CSI volume that corresponds to this volume ID.
```
Nit: ... and snapshot ID? Perhaps ```VolumeIdentifier``` itself is a misnomer since it can represent snapshots? And since it only represents RBD-backed volumes and snapshots, should it be moved to the ```rbd``` package?

ShyamsundarR replied:
Yes, started as a representation for volumes and the name got left behind. Will address naming and related comments in an update to this PR.

I would still like to retain this outside of rbd (i.e in util), as I foresee we would be reusing this in the CephFS plugin code as well, to generate and return Vol|SnapID for volumes from cephfs. objections to retaining these here?

dillaman replied:
Is there a plan for statelessly encoding CephFS volumes? Does it fit here w/o a hammer? 

@batrick

batrick replied:
The term "volume" now is being used to refer to CephFS file systems. It is sufficient to just use the file system cluster id (fscid) as an identifier. Subvolumes are directories within a volume. Subvolumes is usually what end-users will use. Does that answer your question?

We also have the "ceph volume" API. I'm wondering what if any of this should be moved there. I regretfully haven't had much time to work on cleaning it up but one goal was to unify the concept of volumes for CephFS/RBD. This would be useful for simpler administration, presentation, usage monitoring, etc.

dillaman replied:
@batrick So will each CephFS file system map to a single CSI StorageClass -- and then PVCs against that storage class would allocate sub-volumes (directories) within the parent FS (volume)? If that's the case, the ```ImageName``` would represent that sub-directory and could also just be programatically named using a UUID.

gman0 replied:
I'm probably late to the party with such a question, but what about pre-provisioned volumes? The user/operator would need a separate tool to generate the ID in order to mount an image/volume (that has not been provisioned by this driver), right?

batrick replied:
>@batrick So will each CephFS file system map to a single CSI StorageClass -- and then PVCs against that storage class would allocate sub-volumes (directories) within the parent FS (volume)?

I believe so, yes.

> If that's the case, the ImageName would represent that sub-directory and could also just be programatically named using a UUID.

How would you map the UUID to a volume/subvolume?

dillaman replied:
@gman0 AFAIK -- it wouldn't have been supported before. You would have needed to create a dummy PV, delete it behind the scenes, and rename your pre-provisioned image to the CSI image name.

dillaman replied:
@batrick Good question (probably better to discuss in #224). If a Ceph FS ```volume``` maps to a k8s StorageClass, a higher-level tool (e.g. Rook) would be responsible for deploying those I'd suspect. As for how to map a k8s PersistentVolume to a CephFS ```sub-volume```, I guess it depends on what CSI Volume ID would need to encode a link back to the storage class name (to lookup the CephFS ```volume```) and the unique sub-volume name.

ShyamsundarR replied:
This is more a detailing and questions around my (mis)understanding around CephFS and PVCs:

The current options in the [StorageClass](https://github.com/ceph/ceph-csi/blob/csi-v1.0/examples/cephfs/storageclass.yaml) for CephFS has the poolname, monitors, secrets (ID/keys).

A PVC, or a volume from the container orchestrator perspective, is a [sub-directory](https://github.com/ceph/ceph-csi/blob/209774160752ac5665051b2599e9a411496cc388/pkg/cephfs/volume.go#L53) under the (default/singleton?) CephFS instance from CephFS perspective. IOW, we do not have an option to specify different cephfs file system names (like so in the mount `-o mds_namespace=myfs`). Is this not needed? @gman0 any reason to have left this out?

The subdirectory is provided an extended attribute named `ceph.dir.layout.pool`, which is set to the pool name recieved from the StorageClass. The pool name passed in is not reused again, hence need not be encoded in the VolumeID that the plugin returns.

If we start supporting multiple cephfs instances, then we would need to specify the same in the StorageClass and also encode the same in the VolumeID returned. Is this the "file system cluster id (fscid)" being referred to? @dillaman and @batrick is this correct? From where can this fscid be fetched/read to use and what are its size/format specifications?

ajarr replied:
Thalai, if I may...

> 
> A PVC, or a volume from the container orchestrator perspective, is a [sub-directory](https://github.com/ceph/ceph-csi/blob/209774160752ac5665051b2599e9a411496cc388/pkg/cephfs/volume.go#L53) under the (default/singleton?) CephFS instance from CephFS perspective. IOW, we do not have an option to specify different cephfs file system names (like so in the mount `-o mds_namespace=myfs`). Is this not needed? @gman0 any reason to have left this out?
> 

We'll want to add a Storageclass option to specify the filesystem within which the sub-directories/PVs will be created. Currently, sub-directories will be created only within the default filesystem (I think it's filesystem with fscid=1).


 
> The subdirectory is provided an extended attribute named `ceph.dir.layout.pool`, which is set to the pool name recieved from the StorageClass. The pool name passed in is not reused again, hence need not be encoded in the VolumeID that the plugin returns.
> 

FYI, `ceph.dir.layout.pool` xattr is used to specify the pool within which the data of the files belonging to directory will be stored. http://docs.ceph.com/docs/master/cephfs/file-layouts/#file-layouts 


> If we start supporting multiple cephfs instances, then we would need to specify the same in the StorageClass and also encode the same in the VolumeID returned. Is this the "file system cluster id (fscid)" being referred to?

Yes.

> @dillaman and @batrick is this correct? From where can this fscid be fetched/read to use and what are its size/format specifications?

You can fetch it using `ceph fs get <filesystem_name> -f json` (the value of "id" key is the fscid). 
See here,
https://github.com/rook/rook/blob/master/pkg/daemon/ceph/client/filesystem.go#L48

I think it'll be better to specify filesystem's name rather than it's ID in the StorageClass definition.

Do we want to allow the StorageClasses to create filesystems? Or do we want to create them using rook filesystem CRDs before hand
https://github.com/rook/rook/blob/master/Documentation/ceph-filesystem-crd.md ?

ShyamsundarR replied:
> > From where can this fscid be fetched/read to use and what are its size/format specifications?
> 
> You can fetch it using `ceph fs get <filesystem_name> -f json` (the value of "id" key is the fscid).
> See here,
> https://github.com/rook/rook/blob/master/pkg/daemon/ceph/client/filesystem.go#L48

Thanks!

> 
> I think it'll be better to specify filesystem's name rather than it's ID in the StorageClass definition.

In the StorageClass it would be the name (akin to the pool name for RBD backed StorageClasses). In the returned VolID encoded we would use the fscid, so that it is agnostic to renames. (like, we want to store the pool ID in the VolumeID rather than the name, for similar rename constraints). Does this sound fine?

> 
> Do we want to allow the StorageClasses to create filesystems? Or do we expect them to be created before hand, e.g., using rook filesystem CRDs
> https://github.com/rook/rook/blob/master/Documentation/ceph-filesystem-crd.md ?

StorageClass defined filesystem names will not be created by the CSI plugin, it (CSI) expects them to be preexisting. It will only create a sub-directory within the given filesystem for corresponding volume requests. So, it is expected to be created before hand.

ajarr replied:
> In the StorageClass it would be the name (akin to the pool name for RBD backed StorageClasses). In the returned VolID encoded we would use the fscid, so that it is agnostic to renames. (like, we want to store the pool ID in the VolumeID rather than the name, for similar rename constraints). Does this sound fine?

Yes.

dillaman commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,9 +224,430 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	var vi util.VolumeIdentifier
+
+	// generate a uuid for the image name
+	uuid := uuid.NewUUID().String()
```
Nit: same comment here re: package name

ShyamsundarR replied:
Will address in a future update to the PR.

dillaman commented on pkg/util/volid.go:
```diff
@@ -0,0 +1,160 @@
+	[clusterID:36bytes (MAX)] + [-:1byte]
+	[poolID:8bytes] + [-:1byte]
+	[length of ImageName=1:4byte] + [-:1byte]
+	[ImageName: (should not overflow total length of maxVolIDLen)]
```
Nit: really should only be the UUID portion and doesn't need to include the ```csi-vol``` / ```csi-snap``` prefix.

ShyamsundarR replied:
Ack! as I address a similar comment in naming in the rbd package, will address it here.

Just from a read and use perspective it maybe worthwhile to retain the naming, but that can also be construed as lazy programming. Will attempt to see if it makes sense to remove it without impacting readability or adding too much wrapper code.

dillaman commented on pkg/rbd/rbd.go:
```diff
@@ -107,9 +205,7 @@ func (r *Driver) Run(driverName, nodeID, endpoint, configRoot string, containeri
 	r.cd.AddControllerServiceCapabilities([]csi.ControllerServiceCapability_RPC_Type{
 		csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME,
 		csi.ControllerServiceCapability_RPC_PUBLISH_UNPUBLISH_VOLUME,
-		csi.ControllerServiceCapability_RPC_LIST_VOLUMES,
```
Nit: why remove these capabilities?

dillaman replied:
Ack

dillaman commented on pkg/util/cephcmds.go:
```diff
@@ -0,0 +1,364 @@
+
+// RemoveOMap removes the entire omap name passed in and returns ErrOMapNotFound is provided omap
+// is not found in rados
+func RemoveOMap(monitors, adminID, key, poolName, oMapName string) error {
```
Nit: it's really just ```RemoveObject```

ShyamsundarR replied:
ack! will address in a future update to the PR.

dillaman commented on pkg/util/cephcmds.go:
```diff
@@ -0,0 +1,364 @@
+	return nil
+}
+
+// RBDImageInfo strongly typed JSON spec for image info
```
Nit: these remaining things seem like they belong in the ```rbd``` package

ShyamsundarR replied:
I assume this is for the 2 functions `GetSnapInfo` and `GetImageInfo`, would that be right?

My reasoning is that these functions run a Ceph/rbd command and return the results, hence should belong here.

Here are more elaborate reasons/thinking about cephcmds.go and rbd package in general that I have,
1) rbd package functions operate on rbdVol or rbdSnap structures
- Some functions inside rbd package also invoke a corresponding Ceph command (pretty much functions in the file rbd_util.go)
  - This is done based on extracting required information from the rbd[Vol|Snap] structure, but makes the function unusable outside the rbd_util.go context
- Unit testing the rbd_util functions that invoke Ceph commands is not easily feasible

2) cephcmds.go aims to be a pure "invoke said Ceph command" series of functions, and is not tied to rbdVol|Snap structures
- makes these functions in cephcmds.go reusable (when required) and also independently testable
- In the future if any call output/input/parsing needs to change, it can be handled in one place (cephcmds.go)

- Initially was written to avoid constantly extracting required information from rbdVol|Sbap structures (calling getRBDKey in each function say) to make Ceph calls, like in the function reserveRBDVol
- but, over time in the PR made it a wrapper over Ceph commands, in the hope that in the future we would be able to test and update this independently

Ideally I would suggest that we move a bulk of the rbd_util.go code that invokes Ceph commands to this file, and as a result would prefer to retain the following 2 functions here.

Thoughts?

dillaman replied:
I wasn't concerned about the generic ```ceph``` CLI commands (beyond the usual brittleness of relying on CLI commands vice a testable API). The mixing of RBD-unique data types, to me, is polluting the diving line between the two. Perhaps the ```rbd``` package just needs an equivalent ```rbdcmds``` file?

ShyamsundarR replied:
Not attempting to press a point, as I do not have any objections to move the functions to the rbd_util.go file (as on further thinking, testing the functions in either package should be feasible and nothing special), BUT,
If the structures were named ImageInfo or some such and not RBD prefixed would that have changed your mind?

I ask as the RBD prefix naming of these was more from the rbd CLI usage and return values, than the rbd golang package usage in this repository, and hence was curious.

ShyamsundarR replied:
Moved rbd specifc CLI calls to the rbd package.

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -205,9 +224,430 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+}
+
+/*
+reserverRBDSnap and reserveRBDVol add respective entries to the volumes and snapshots omaps, post
```
Will address in a future update to the PR.

ShyamsundarR commented on pkg/rbd/rbd.go:
```diff
@@ -27,10 +27,89 @@ import (
+	snapImgOMap = "csi.snap."
+	// key in snapImgOMap containing CSI snapshot name for which the snapshot was created
+	snapImgOMapNameKey = "csi.snapname"
+	// key in snapImgOMap containing RBD source image name for which the snapshot was created
```
This [stores](https://github.com/ceph/ceph-csi/pull/288/files/0778e05af7b10390a2738ae57d051b5072da2efe#diff-8e27516879efe971bf4755fadf585e06R562) the RBD image name that is the source of this snapshot.

ShyamsundarR commented on pkg/rbd/rbd.go:
```diff
@@ -27,10 +27,89 @@ import (
+RBD image can satisfy the requirements for the request
+  - If during the process of checking the same, it is found that some linking information is stale
+  or missing, the corresponding keys upto the key in the volOMap is cleaned up, to start afresh
+- If the key is not found, or was cleaned up, the request is treated as a fresh create request, and
```
I am unclear which part this review comment refers to, but assuming the following and responding.

The comment is to do the following,
- `rados [...] create csi.volume.<image-uuid>`
  - Check for EEXISTS, to deal with image-uuid in use, and handle failure paths
- On success,
  - `rados [...] setomapval csi.volume.<image-uuid> csi.volname <volume-name>`

Instead of the current method of using just setomapval, to create the omap and set the key and value, in one shot.

-----------------------------

Ack, just tried using setomapval on an existing omap and key, and as expected it overwrites the value in the key, which is not what we require. Creating the omap first (and dealing with exists situations) and then setting the key is the safe thing to do.

Will address in an update to this PR.

ShyamsundarR commented:
___
*[pkg/rbd/rbd_util.go, line 271 at r1](https://reviewable.io/reviews/ceph/ceph-csi/288#-LbDjq0gj9WhBXK2MXJ3-r1-271:-LbTHe5d7z3IrV-DpZlq:b-sdnyrl) ([raw file](https://github.com/ceph/ceph-csi/blob/0778e05af7b10390a2738ae57d051b5072da2efe/pkg/rbd/rbd_util.go#L271)):*
<details><summary><i>Previously, dillaman (Jason Dillaman) wrote…</i></summary><blockquote>

Fair enough. Long term efficiency, we really should just get the entire CSI off using the CLI (and associated parsing of CLI output and error codes) and instead directly use the golang wrapper API [1]

[1] https://github.com/ceph/go-ceph
</blockquote></details>

Agreed, we should move to that at some point in time in the future.


<!-- Sent from Reviewable.io -->

ShyamsundarR commented:
@dillaman @Madhu-1 @gman0 Addressed review comments in a new commit, please review.

IMO, the code is complete, it does need a fresh PR against master now, but it also needs the deployment and documentation changes to be made. As I get those done, request a review of the code as a part of this PR itself, so that I can address any concerns in the next round of updates.

Items that are unresolved or not going to be resolved in this PR IMO are,
- Naming of the PoolID field in CsiIdentifier structure in file volid.go file (thinking about that now, the file should be renamed csiid.go :-/ ) This will be done when we move to making cephfs stateless

- 1:1 clusterID secret to storage class
  - This is a broader change, and IMO encoding the pool ID in the returned VolID, does not preclude moving the pool name to the secret in the future
  - Hence, for now, retaining the existing method from the in-tree and CSI implementation of specifying the pool name in the StorageClass

ShyamsundarR commented:
This is closed, as the PR is moved to master https://github.com/ceph/ceph-csi/pull/312 please continue the reviews there.
